addon/services/sticky-manager.js

```diff
@@ -0,0 +1,245 @@
+      let placeHolderNodeCurrentTop =
+        getRectForNode(placeHolderNode).top - containerTop + scrollTop;
+      let top = containerTop + stickyOffset;
+      let width = getRectForNode(placeHolderNode).width;
+      let left = getRectForNode(placeHolderNode).left;
+
+      let sticky;
+
+      if (boundingElement == null) {
+        sticky = scrollPosition >= placeHolderNodeCurrentTop;
+      } else {
+        let stickyItemHeight = getRectForNode(stickyNode).height;
+        let stickyItemBottomPosition =
+          getRectForNode(boundingElement).bottom -
+          stickyItemHeight +
+          scrollTop -
+          containerTop;
+
+        sticky =
+          scrollPosition >= placeHolderNodeCurrentTop &&
+          scrollPosition < stickyItemBottomPosition;
+      }
+
+      return {
+        sticky,
+        top,
+        left,
+        width,
+      };
+    },
+
+    updateStuckItems(item, sticky) {
+      let { stickyNode } = item;
+      if (sticky && !this.isNodeStuck(stickyNode)) {
+        this.addStuckItem(item);
+      } else if (!sticky && this.isNodeStuck(stickyNode)) {
+        this.removeStuckItem(item);
+      }
+    },
+
+    addStuckItem(stickyItem) {
+      this.get('stuckItems').push(stickyItem);
+    },
+
+    removeStuckItem(stickyItem) {
+      let stuckItems = this.get('stuckItems');
+      let { stickyNode: nodeToRemove } = stickyItem;
+      let nodeIndex = stuckItems.findIndex(
+        ({ stickyNode }) => nodeToRemove === stickyNode
+      );
+      stuckItems.splice(nodeIndex, 1);
+    },
+
+    getOffset(node) {
+      let stuckItems = this.get('stuckItems');
+      let stuckItemsLength = stuckItems.get('length');
+      if (stuckItemsLength === 0) {
+        return 0;
+      }
+
+      let offset = 0;
+      let count = 0;
+      let stuckNodesLength = stuckItemsLength;
+      let nodeRect = getRectForNode(node);
+
+      while (count < stuckNodesLength) {
+        let stuckNode = stuckItems[count].stickyNode;
+        if (stuckNode !== node) {
+          let stuckNodeRect = getRectForNode(stuckNode);
+          if (!horizontallyOverlaps(nodeRect, stuckNodeRect)) {
+            offset += getRectForNode(stuckNode).height;
+          }
+        } else {
+          break;
+        }
+        count++;
+      }
+
+      return offset;
+    },
+
+    isNodeStuck(node) {
+      let nodeFound = this.get('stuckItems').findIndex(
+        ({ stickyNode }) => node === stickyNode
+      );
+
+      return nodeFound >= 0;
+    },
+
+    init() {
+      this._super(...arguments);
+
+      let container = this.get('container') || document;
+      if (container) {
+        this.setContainer(container);
+      }
+    },
+  }
+);
+
+function isDocument(node) {
+  return node === document;
+}
+
+function scrollTopFor(container) {
+  return isDocument(container)
+    ? document.body.scrollTop || document.documentElement.scrollTop
+    : container.scrollTop;
+}
+
+function horizontallyOverlaps(rect1, rect2) {
```

Comment:
assuming you copied this over from the react implementation? How do you feel about renaming them with string prefixes as `firstRect` and `secondRect`? probably reduces chance for typos

Reply:
I'd like to keep the naming the same so future diffs with the React code are clearer